aiida/engine/daemon/execmanager.py

```diff
@@ -159,6 +159,13 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
     remote_symlink_list = calc_info.remote_symlink_list or []
     provenance_exclude_list = calc_info.provenance_exclude_list or []
 
+    # Creates the directory structure locally and first to prevent unnecessary transport
```

Comment:
Maybe for this comment, more than explaining that this is a solution to a problem that never was in the code (which anyway you could keep, but maybe saying 'if we were to create folders only on the remote, for each file), you could add an explanation of the original problem you wanted to solve (if the copy_list include only the filename of a file in a subfolder, and the subfolder does not exist, they would fail - so we first create possibly empty folders where appropriate.

Reply:
Changed it for the following text; hopefully this is clearer now, if you don't object I'll merge when the tests pass.

```
    # First creates the directory structure locally before copying the sandbox folder, so that all the intermediate
    # folders for the files in the copy_lists are there before calling the copy methods of the transport (or else
    # these will fail).
    # Alternatively, one would have to call the path creation methods of the transports just before calling the
    # copy methods to make sure each path is there, unnecessarily duplicating the number of connections requested.
```

Reply:
Text is fine, just a minor nitpick: it is not so much requesting new connections each time, it already has an open connection. But each command *is* calling a command over the transport, which has a certain base delay that can be non-negligible.